Trace scattered tensor options arguments

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44158 VariableKernel calls into scattered C++ api
* #44087 Add faithful C++ API
* **#44071 Trace scattered tensor options arguments**
* #44062 Remove hacky_wrapper from BackendSelect kernels
* #44005 Remove hacky_wrapper from VariableType and TraceType

Previously, tracing re-gathered ScalarType, Layout, Device, bool into a TensorOptions object and called `tracer::addInput()` on the gathered TensorOptions argument. `tracer::addInput()` then scattered them again and added the individual scattered arguments to the traced graph. This PR avoids the extraneous gathering and re-scattering step and calls `tracer::addInput()` on the individual arguments directly. This avoid the perf hit for an unnecessary gathering step.

This applies to both c10-full and non-c10-full ops. In the case of c10-full ops, the tracing kernels takes scattered arguments and we can directly pass them to `tracer::addInput()`. In the case of non-c10-full ops, the kernel takes a `TensorOptions` argument but we still call `tracer::addInput()` on the scattered arguments.

Differential Revision: [D23486638](https://our.internmc.facebook.com/intern/diff/D23486638/)

smessmer committed:
Trace scattered tensor options arguments

Previously, tracing re-gathered ScalarType, Layout, Device, bool into a TensorOptions object and traced them as one TensorOptions argument.
Now, we're tracing the individual arguments.

This avoid the perf hit for an unnecessary gathering step, and also makes more sense semantically.

Differential Revision: [D23486638](https://our.internmc.facebook.com/intern/diff/D23486638/)

[ghstack-poisoned]

bhosmer reviewed (approved):
Couple minor things - also though, at least some of the test failures look like they might be related

smessmer committed:
Update on "Trace scattered tensor options arguments"


Previously, tracing re-gathered ScalarType, Layout, Device, bool into a TensorOptions object and called `tracer::addInput()` on the gathered TensorOptions argument. `tracer::addInput()` then scattered them again and added the individual scattered arguments to the traced graph. This PR avoids the extraneous gathering and re-scattering step and calls `tracer::addInput()` on the individual arguments directly. This avoid the perf hit for an unnecessary gathering step.

This applies to both c10-full and non-c10-full ops. In the case of c10-full ops, the tracing kernels takes scattered arguments and we can directly pass them to `tracer::addInput()`. In the case of non-c10-full ops, the kernel takes a `TensorOptions` argument but we still call `tracer::addInput()` on the scattered arguments.

Differential Revision: [D23486638](https://our.internmc.facebook.com/intern/diff/D23486638/)

[ghstack-poisoned]

smessmer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -71,12 +71,18 @@
 RENAME_TRACE_ADD_ARGS = {
     'fill': '''\
-    jit::tracer::addInputs(node, "options", TensorOptions());
+    jit::tracer::addInputs(node, "dtype", c10::optional<ScalarType>());
```
good point, I should change it to the default instead. Not sure what's the right thing to do but I want to avoid behavior changes for now.

smessmer replied:
ugh turns out this actually used `TensorOptions::dtype_opt()`, so it would in fact trace a `nullopt`. But `layout, device, pin_memory` used default values. This is almost certainly wrong but I'm replicating the logic for now.

smessmer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -494,39 +500,40 @@ def format_trace_op_name(declaration):
+        trace_input_spec = [(i['name'], i['name'], i['type'], i.get('is_nullable')) for i in trace_inputs]
+    else:
+        trace_inputs = declaration['arguments']
+        trace_input_spec = [(i['name'], i['name'], i['simple_type'], i.get('is_nullable')) for i in trace_inputs]
```
I don't know much about the difference between `simple_type` and `type`. `simple_type` seems to be only present for non-c10-full ops, but for them it has to be used instead of `type`, otherwise it doesn't compile.

smessmer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -494,39 +500,40 @@ def format_trace_op_name(declaration):
+                result += ADD_TRACE_INPUT.substitute(name="dtype", input="optTypeMetaToScalarType(options.dtype_opt())") + "\n"
+                result += ADD_TRACE_INPUT.substitute(name="layout", input="options.layout_opt()") + "\n"
+                result += ADD_TRACE_INPUT.substitute(name="device", input="options.device_opt()") + "\n"
+                result += ADD_TRACE_INPUT.substitute(name="pin_memory", input="options.pinned_memory_opt()")
```
there's some differences between them if you look at the details. For example, they have different argument names, and this one here uses `options` while the one below uses `out.options`. I could try and reuse logic there, but that logic would have to be generic enough to support these differences, which kills its benefits in my opinion.

smessmer committed:
Update on "Trace scattered tensor options arguments"


Previously, tracing re-gathered ScalarType, Layout, Device, bool into a TensorOptions object and called `tracer::addInput()` on the gathered TensorOptions argument. `tracer::addInput()` then scattered them again and added the individual scattered arguments to the traced graph. This PR avoids the extraneous gathering and re-scattering step and calls `tracer::addInput()` on the individual arguments directly. This avoid the perf hit for an unnecessary gathering step.

This applies to both c10-full and non-c10-full ops. In the case of c10-full ops, the tracing kernels takes scattered arguments and we can directly pass them to `tracer::addInput()`. In the case of non-c10-full ops, the kernel takes a `TensorOptions` argument but we still call `tracer::addInput()` on the scattered arguments.

Differential Revision: [D23486638](https://our.internmc.facebook.com/intern/diff/D23486638/)

[ghstack-poisoned]

smessmer committed:
Update on "Trace scattered tensor options arguments"


Previously, tracing re-gathered ScalarType, Layout, Device, bool into a TensorOptions object and called `tracer::addInput()` on the gathered TensorOptions argument. `tracer::addInput()` then scattered them again and added the individual scattered arguments to the traced graph. This PR avoids the extraneous gathering and re-scattering step and calls `tracer::addInput()` on the individual arguments directly. This avoid the perf hit for an unnecessary gathering step.

This applies to both c10-full and non-c10-full ops. In the case of c10-full ops, the tracing kernels takes scattered arguments and we can directly pass them to `tracer::addInput()`. In the case of non-c10-full ops, the kernel takes a `TensorOptions` argument but we still call `tracer::addInput()` on the scattered arguments.

Differential Revision: [D23486638](https://our.internmc.facebook.com/intern/diff/D23486638/)

[ghstack-poisoned]

smessmer committed:
Update on "Trace scattered tensor options arguments"


Previously, tracing re-gathered ScalarType, Layout, Device, bool into a TensorOptions object and called `tracer::addInput()` on the gathered TensorOptions argument. `tracer::addInput()` then scattered them again and added the individual scattered arguments to the traced graph. This PR avoids the extraneous gathering and re-scattering step and calls `tracer::addInput()` on the individual arguments directly. This avoid the perf hit for an unnecessary gathering step.

This applies to both c10-full and non-c10-full ops. In the case of c10-full ops, the tracing kernels takes scattered arguments and we can directly pass them to `tracer::addInput()`. In the case of non-c10-full ops, the kernel takes a `TensorOptions` argument but we still call `tracer::addInput()` on the scattered arguments.

Differential Revision: [D23486638](https://our.internmc.facebook.com/intern/diff/D23486638/)

[ghstack-poisoned]

smessmer committed:
Update on "Trace scattered tensor options arguments"


Previously, tracing re-gathered ScalarType, Layout, Device, bool into a TensorOptions object and called `tracer::addInput()` on the gathered TensorOptions argument. `tracer::addInput()` then scattered them again and added the individual scattered arguments to the traced graph. This PR avoids the extraneous gathering and re-scattering step and calls `tracer::addInput()` on the individual arguments directly. This avoid the perf hit for an unnecessary gathering step.

This applies to both c10-full and non-c10-full ops. In the case of c10-full ops, the tracing kernels takes scattered arguments and we can directly pass them to `tracer::addInput()`. In the case of non-c10-full ops, the kernel takes a `TensorOptions` argument but we still call `tracer::addInput()` on the scattered arguments.

Differential Revision: [D23486638](https://our.internmc.facebook.com/intern/diff/D23486638/)

[ghstack-poisoned]

smessmer committed:
Update on "Trace scattered tensor options arguments"


Previously, tracing re-gathered ScalarType, Layout, Device, bool into a TensorOptions object and called `tracer::addInput()` on the gathered TensorOptions argument. `tracer::addInput()` then scattered them again and added the individual scattered arguments to the traced graph. This PR avoids the extraneous gathering and re-scattering step and calls `tracer::addInput()` on the individual arguments directly. This avoid the perf hit for an unnecessary gathering step.

This applies to both c10-full and non-c10-full ops. In the case of c10-full ops, the tracing kernels takes scattered arguments and we can directly pass them to `tracer::addInput()`. In the case of non-c10-full ops, the kernel takes a `TensorOptions` argument but we still call `tracer::addInput()` on the scattered arguments.

Differential Revision: [D23486638](https://our.internmc.facebook.com/intern/diff/D23486638/)

[ghstack-poisoned]

smessmer committed:
Update on "Trace scattered tensor options arguments"


Previously, tracing re-gathered ScalarType, Layout, Device, bool into a TensorOptions object and called `tracer::addInput()` on the gathered TensorOptions argument. `tracer::addInput()` then scattered them again and added the individual scattered arguments to the traced graph. This PR avoids the extraneous gathering and re-scattering step and calls `tracer::addInput()` on the individual arguments directly. This avoid the perf hit for an unnecessary gathering step.

This applies to both c10-full and non-c10-full ops. In the case of c10-full ops, the tracing kernels takes scattered arguments and we can directly pass them to `tracer::addInput()`. In the case of non-c10-full ops, the kernel takes a `TensorOptions` argument but we still call `tracer::addInput()` on the scattered arguments.

Differential Revision: [D23486638](https://our.internmc.facebook.com/intern/diff/D23486638/)

[ghstack-poisoned]

smessmer committed:
Update on "Trace scattered tensor options arguments"


Previously, tracing re-gathered ScalarType, Layout, Device, bool into a TensorOptions object and called `tracer::addInput()` on the gathered TensorOptions argument. `tracer::addInput()` then scattered them again and added the individual scattered arguments to the traced graph. This PR avoids the extraneous gathering and re-scattering step and calls `tracer::addInput()` on the individual arguments directly. This avoid the perf hit for an unnecessary gathering step.

This applies to both c10-full and non-c10-full ops. In the case of c10-full ops, the tracing kernels takes scattered arguments and we can directly pass them to `tracer::addInput()`. In the case of non-c10-full ops, the kernel takes a `TensorOptions` argument but we still call `tracer::addInput()` on the scattered arguments.

Differential Revision: [D23486638](https://our.internmc.facebook.com/intern/diff/D23486638/)

[ghstack-poisoned]
